examples/transformer_base.py

```diff
@@ -0,0 +1,279 @@
+
+        super(BaseTransformer, self).__init__()
+        self.hparams = hparams
+        args = self.hparams
+        args.model_type = args.model_type.lower()
+
+        config_class, model_class, tokenizer_class = MODEL_CLASSES[args.model_type]
+        config = config_class.from_pretrained(
+            args.config_name if args.config_name else args.model_name_or_path,
+            num_labels=num_labels,
+            cache_dir=args.cache_dir if args.cache_dir else None,
+        )
+        tokenizer = tokenizer_class.from_pretrained(
+            args.tokenizer_name if args.tokenizer_name else args.model_name_or_path,
+            do_lower_case=args.do_lower_case,
+            cache_dir=args.cache_dir if args.cache_dir else None,
+        )
+        model = model_class.from_pretrained(
+            args.model_name_or_path,
+            from_tf=bool(".ckpt" in args.model_name_or_path),
+            config=config,
+            cache_dir=args.cache_dir if args.cache_dir else None,
+        )
+        self.config, self.tokenizer, self.model = config, tokenizer, model
+
+
+    def configure_optimizers(self):
+        "Prepare optimizer and schedule (linear warmup and decay)"
+        args = self.hparams
+        model = self.model
+
+        t_total = len(self.train_dataloader()) // args.gradient_accumulation_steps * float(args.num_train_epochs)
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in model.named_parameters()
+                           if not any(nd in n for nd in no_decay)],
+                "weight_decay": args.weight_decay,
+            },
+            {"params": [p for n, p in model.named_parameters()
+                        if any(nd in n for nd in no_decay)],
+             "weight_decay": 0.0},
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters,
+                          lr=args.learning_rate,
+                          eps=args.adam_epsilon)
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer,
+            num_warmup_steps=args.warmup_steps,
+            num_training_steps=t_total
+        )
+        self.lr_scheduler = scheduler
+        return [optimizer]
+
+
+    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
+                       second_order_closure=None):
+
+        # Step each time.
+        optimizer.step()
+        optimizer.zero_grad()
+        self.lr_scheduler.step()
+
+    def get_tqdm_dict(self):
+        tqdm_dict = {
+            'loss': '{:.3f}'.format(self.trainer.avg_loss),
+            'lr' : self.lr_scheduler.get_last_lr()[-1]
+        }
+
+        return tqdm_dict
+
+    def test_step(self, batch, batch_nb):
+        return self.validation_step(batch, batch_nb)
+
+    def test_end(self, outputs):
+        return self.validation_end(outputs)
+
+    @pl.data_loader
+    def train_dataloader(self):
+        args = self.hparams
```

Comment:
(nit) I don't understand purpose of this line.

Reply:
I'll switch to hparams. (old code had args)